### PR TITLE
[mqtt] Retry non-iothub subscription when rejected by server

### PR DESCRIPTION
**Pull request by dmolokanov**
When edge deployed on several levels in nested scenarios there can be a lag between all rules are applied. In this case, lower level subscriptions can be rejected by the upper level. The lower level should keep retrying until it is subscribed.

### CODE REVIEW

File: mqtt/mqtt-bridge/src/pump/builder.rs [hunk 1]
```diff
@@ -128,10 +129,14 @@ where
...
         let remote_topic_mappers_updates = TopicMapperUpdates::new(topic_filters);
 
+        let (retry_send, retry_recv) = mpsc::unbounded_channel();
+
         let rpc_subscriptions = RpcSubscriptions::default();
         let rpc = RemoteRpcMqttEventHandler::new(rpc_subscriptions.clone(), local_pump.handle());
-        let messages =
+        let mut messages =
             StoreMqttEventHandler::new(local_store, remote_topic_mappers_updates.clone());
+        messages.set_retry_sub_sender(retry_send);
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-bridge/src/pump/builder.rs:138]: nit: is it possible to just add retry_sender to ctor?

File: mqtt/mqtt-bridge/src/pump/mod.rs [hunk 2]
```diff
@@ -16,7 +16,7 @@ use futures_util::{
     join, pin_mut,
 };
 use mockall::automock;
-use parking_lot::Mutex;
+use parking_lot::RwLock;
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-bridge/src/pump/mod.rs:19]: just curious why this change in this PR and not separate?

**Review by vadim-kovalyov — APPROVED** [review]
LGTM!

File: mqtt/mqtt-bridge/src/pump/mod.rs (hunk 2 above)
Comment (dmolokanov) [reply mqtt/mqtt-bridge/src/pump/mod.rs:19]: I started to use more read operations with changes in this PR: reading when retry timeout expired. so they can come together

File: mqtt/mqtt-bridge/src/pump/builder.rs (hunk 1 above)
Comment (dmolokanov) [reply mqtt/mqtt-bridge/src/pump/builder.rs:138]: not every pump needs to retry subscriptions thus `retry_send` can be optional. not sure if optional argument is good for ctor

Comment (vadim-kovalyov) [reply mqtt/mqtt-bridge/src/pump/builder.rs:138]: I see. The ideal solution would be builder :) but maybe overkill here. So i'm fine with "setter"

File: mqtt/mqtt-bridge/src/messages.rs [hunk 3]
```diff
@@ -76,19 +98,28 @@ impl<S> StoreMqttEventHandler<S> {
...
         if self.topic_mappers.remove(sub).is_none() {
             warn!("unexpected subscription/rejected ack for {}", sub);
         };
     }
+
+    fn handle_rejected(&mut self, sub: SubscribeTo) {
+        self.topic_mappers.remove(&sub.topic_filter);
+        if let Some(sender) = &mut self.retry_sub_send {
+            if sender.send(sub).is_err() {
+                warn!("unable to schedule subscription retry. channel closed");
```
Comment (nyanzebra) [inline mqtt/mqtt-bridge/src/messages.rs:119]: Can we log the err too?

File: mqtt/mqtt-bridge/src/pump/builder.rs (hunk 1 above)
Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/pump/builder.rs:138]: There is a macro trait for builder pattern :) https://github.com/colin-kiegel/rust-derive-builder

File: mqtt/mqtt-bridge/src/messages.rs [hunk 4]
```diff
@@ -150,6 +181,32 @@ where
     }
 }
 
+pub async fn retry_subscriptions(
+    retries: UnboundedReceiver<SubscribeTo>,
```
Comment (nyanzebra) [inline mqtt/mqtt-bridge/src/messages.rs:185]: Why are we choosing unbounded?

File: mqtt/mqtt-bridge/src/messages.rs (hunk 3 above)
Comment (dmolokanov) [reply mqtt/mqtt-bridge/src/messages.rs:119]: in this case the error is the `sub` itself. the only way error occurred is when the channel is closed https://docs.rs/tokio/1.0.2/tokio/sync/mpsc/struct.UnboundedSender.html#method.send

File: mqtt/mqtt-bridge/src/pump/builder.rs (hunk 1 above)
Comment (dmolokanov) [reply mqtt/mqtt-bridge/src/pump/builder.rs:138]: well, some time ago we decided not to use too many derive macros: it makes code more difficult to read and navigate and it is super easy to implement it manually.

we write code once and read it many times...

File: mqtt/mqtt-bridge/src/messages.rs (hunk 4 above)
Comment (dmolokanov) [reply mqtt/mqtt-bridge/src/messages.rs:185]: no specific reason. we could use either of the channels in this case

Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/messages.rs:185]: Are we expecting a lot of simultaneous retries? As unbounded will take up more memory:

```
Note that the amount of available system memory is an implicit bound to the channel. Using an unbounded channel has the ability of causing the process to run out of memory. In this case, the process will be aborted.
```
